File: jar-shell/src/main/java/com/beancounter/shell/cli/UserCommands.java
package com.beancounter.shell.cli;

import com.beancounter.auth.client.LoginService;
import com.beancounter.client.services.RegistrationService;
import com.beancounter.common.contracts.RegistrationRequest;
import com.beancounter.common.exception.UnauthorizedException;
import com.beancounter.common.utils.BcJson;
import com.beancounter.shell.config.EnvConfig;
import com.fasterxml.jackson.core.JsonProcessingException;
import lombok.SneakyThrows;
import org.jline.reader.LineReader;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.shell.standard.ShellComponent;
import org.springframework.shell.standard.ShellMethod;
import org.springframework.shell.standard.ShellOption;

@ShellComponent
public class UserCommands {

  private final LoginService loginService;
  private final RegistrationService registrationService;
  private final EnvConfig envConfig;
  private LineReader lineReader;

  public UserCommands(LoginService loginService,
                      RegistrationService registrationService,
                      EnvConfig envConfig) {
    this.loginService = loginService;
    this.registrationService = registrationService;
    this.envConfig = envConfig;
  }

  @Autowired
  public void setLineReader(@Lazy LineReader lineReader) {
    this.lineReader = lineReader;
  }

  @ShellMethod("Identify yourself")
  public void login(
      @ShellOption(help = "User ID") String user) {

    String password = lineReader.readLine("Password: ", '*');
    this.loginService.login(user, password, envConfig.getClient());

  }

  @ShellMethod("What's my access token?")
  public String token() {
    return registrationService.getToken();
  }

  @ShellMethod("Who am I?")
  @SneakyThrows
  public String me() {
    return BcJson.getWriter().writeValueAsString(this.registrationService.me());
  }

  @ShellMethod("Register your Account")
  public String register() throws JsonProcessingException {
    JwtAuthenticationToken token = registrationService.getJwtToken();
    if (token == null) {
      throw new UnauthorizedException("Please login");
    }
    return BcJson.getWriter()
        .writeValueAsString(
            this.registrationService
                .register(new RegistrationRequest(token.getToken().getClaim("email")))
        );


  }

}

File: jar-shell/src/main/kotlin/com/beancounter/shell/cli/UserCommands.kt
package com.beancounter.shell.cli

import com.beancounter.auth.client.LoginService
import com.beancounter.client.services.RegistrationService
import com.beancounter.common.contracts.RegistrationRequest
import com.beancounter.common.utils.BcJson
import com.beancounter.shell.config.EnvConfig
import com.fasterxml.jackson.core.JsonProcessingException
import org.jline.reader.LineReader
import org.springframework.context.annotation.DependsOn
import org.springframework.shell.standard.ShellComponent
import org.springframework.shell.standard.ShellMethod
import org.springframework.shell.standard.ShellOption

/**
 * Interactive access to BeanCounter from the commandline.
 */
@ShellComponent
@DependsOn("shell", "lineReader")
class UserCommands(
    private val loginService: LoginService,
    private val registrationService: RegistrationService,
    private val envConfig: EnvConfig,
    private val lineReader: LineReader,
) {
    private val bcJson = BcJson()

    @ShellMethod("Identify yourself")
    fun login(
        @ShellOption(help = "User ID") user: String,
    ) {
        val password = lineReader.readLine("Password: ", '*')
        loginService.login(user, password, envConfig.client)
    }

    @ShellMethod("What's my access token?")
    fun token(): String {
        return if (registrationService.token == null) {
            "Not logged in"
        } else {
            registrationService.token!!
        }
    }

    @ShellMethod("Who am I?")
    fun me(): String {
        return bcJson.writer.writeValueAsString(registrationService.me())
    }

    @ShellMethod("Register your Account")
    @Throws(JsonProcessingException::class)
    fun register(emailClaim: String): String {
        val token = registrationService.jwtToken
        return bcJson.writer
            .writeValueAsString(
                registrationService
                    .register(RegistrationRequest(token.token.getClaim(emailClaim))),
            )
    }
}
